File: python/79_Word_Search.py
#!/usr/bin/python

class Solution(object):
    def exist(self, board, word):    	
    	m, n = len(board), len(board[0])
    	if not m or not n:
    		return False
    	track = [[0] * n for i in range(m)]
    	for i in range(m):
    		for j in range(n):
    			if self.dfs(board, i, j, track, word, 0, m, n):
    				return True
    	return False

    def dfs(self, board, i, j, track, word, index, m, n):
    	if index == len(word):
    		return True
    	if i < 0 or i >= m or j < 0 or j >= n or board[i][j] != word[index] or track[i][j]:
    		return False
    	track[i][j] = 1
    	# important, if only one of 4 directions works, then match
    	if self.dfs(board, i + 1, j, track, word, index + 1, m, n): return 1
    	if self.dfs(board, i - 1, j, track, word, index + 1, m, n): return 1
    	if self.dfs(board, i, j + 1, track, word, index + 1, m, n): return 1
    	if self.dfs(board, i, j - 1, track, word, index + 1, m, n): return 1
    	track[i][j] = 0 # important, recover tag when we haven't tried that route.
    	return False


def main():
    s = Solution()
    dat = ["CAA",
           "AAA",
           "BCD"]
    print s.exist(dat, "AAB")


if __name__ == "__main__":
    main()

File: c++/79_Word_Search.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    bool exist(vector<vector<char> >& board, string word) {
    	if (board.size() == 0 || board[0].size() == 0)
    		return false;
    	int m = board.size(), n = board[0].size();
    	vector<vector<bool> > track(m, vector<bool> (n, false));
    	for (int i = 0; i < m; i++) {
    		for (int j = 0; j < n; j++) {
    			if (dfs(board, i, j, m, n, 0, word, track))
    				return true;
    		}
    	}
    	return false;
    }

    bool dfs(vector<vector<char> >& board, int i, int j, int m, int n, int index, string word, vector<vector<bool> > &track) {
    	if (index == word.size())
    		return true;
    	if (i < 0 || i >= m || j < 0 || j >= n || board[i][j] != word[index] || track[i][j])
    		return false;
    	track[i][j] = 1;
    	// classical part
    	if (dfs(board, i + 1, j, m, n, index + 1, word, track)) return true;
    	if (dfs(board, i - 1, j, m, n, index + 1, word, track)) return true;
    	if (dfs(board, i, j + 1, m, n, index + 1, word, track)) return true;
    	if (dfs(board, i, j - 1, m, n, index + 1, word, track)) return true;
    	track[i][j] = 0; // hard part
    	return false;
    }
};


int main() {
	Solution s;
	vector<vector<char> > board;
	string word;
	bool tag =s.exist(board, word);
}